### PR TITLE
feat(connector): [Fiserv] add Refunds, Cancel and Wallets flow along with Unit Tests

### PR DESCRIPTION
**Pull request by SamraatBansal**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
add refunds, wallets, cancel and capture flows for Fiserv

### Additional Changes

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Requires the implementation of Refunds flow and Cancel flow along with wallets.

## How did you test it?

The 2 Void tests were ignored as they require a different merchant ID, which is created by default in the Fiserv Dashboard, to test void flows use those MID

Screenshot 2023-02-11 at 2 06 20 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [X] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/fiserv.rs [hunk 1]
```diff
@@ -108,6 +238,104 @@ impl
...
+                headers::CONTENT_TYPE.to_string(),
+                types::PaymentsAuthorizeType::get_content_type(self).to_string(),
+            ),
+            ("Client-Request-Id".to_string(), client_request_id),
+            ("Auth-Token-Type".to_string(), "HMAC".to_string()),
+            ("Api-Key".to_string(), api_key),
+            ("Timestamp".to_string(), timestamp.to_string()),
+            ("Authorization".to_string(), hmac),
+        ];
+        Ok(headers)
```
Comment (Narayanbhat166) [inline crates/router/src/connector/fiserv.rs:269]: Repeated code, utilize a common function for the common headers, and mutate the headers for specific requests.

File: crates/router/src/connector/fiserv.rs [hunk 2]
```diff
@@ -108,6 +238,104 @@ impl
...
 {
+    fn get_headers(
+        &self,
+        req: &types::PaymentsSyncRouterData,
+        _connectors: &settings::Connectors,
+    ) -> CustomResult<Vec<(String, String)>, errors::ConnectorError> {
+        let timestamp = OffsetDateTime::now_utc().unix_timestamp_nanos() / 1_000_000;
+        let auth: fiserv::FiservAuthType =
+            fiserv::FiservAuthType::try_from(&req.connector_auth_type)?;
+        let api_key = auth.api_key.clone();
```
Comment (Narayanbhat166) [inline crates/router/src/connector/fiserv.rs:249]: This logic can be added to the get_auth_header function of connector common

File: crates/router/src/connector/fiserv.rs [hunk 3]
```diff
@@ -108,6 +238,104 @@ impl
...
+        let headers = vec![
+            (
+                headers::CONTENT_TYPE.to_string(),
+                types::PaymentsAuthorizeType::get_content_type(self).to_string(),
+            ),
+            ("Client-Request-Id".to_string(), client_request_id),
+            ("Auth-Token-Type".to_string(), "HMAC".to_string()),
+            ("Api-Key".to_string(), api_key),
+            ("Timestamp".to_string(), timestamp.to_string()),
+            ("Authorization".to_string(), hmac),
```
Comment (Narayanbhat166) [inline crates/router/src/connector/fiserv.rs:267]: please try to use the constants defined in header module, this will help in reducing typos

File: crates/router/src/connector/fiserv/transformers.rs [hunk 4]
```diff
@@ -64,60 +90,74 @@ pub struct TransactionInteraction {
...
+                // println!("oooooooooo{:?}", json_value);
+                match wallet_data.issuer_name{
+                    api_models::enums::WalletIssuer::GooglePay => Source::GooglePay {
+                        data: r#"{\"encryptedMessage\":\"XqpMeRFRqQJkusxA/yCsxLHwW/KaSWGAEjqoo0uinoLGbzZJ3oMqir/Ldq+FhSSDHH+jNfIrT0Zfv32F0LzjxRgjjUiu6bhgBa4376JEhtH7NmrEb76pS0KZiqdIqghtpUY0eLMkVz4bpGghsr7SFAEPyjzvHxSSRAik3ZCM5vKDuA4pGCsH6zCYr12JGUafB6Mvbauu3+GVsIcJhnG/j6qr1Jos8b1V5ywEyY1Xgs9fyD2IwGQ04ni7BqVqYNyPzqTEXNcRVnW7nzk1KXBhcT/S2Ug36rax0hDNUQxAS09gqw4Xof6ixLcXCpXhgSGNje3l1AFdWu1chRI6fVZe77mgR66FMAVCdLYEQmouUIBtBCT5WCQKPxxg4Hiu/DU3bhOw09ZYT6v8n0U4ai5wJOj2QwqRHEaGUhBN+2WuJofhxDixprdyVdLdmkuxdKzLyQ\\u003d\\u003d\",\"ephemeralPublicKey\":\"BIfDllJB8/HGArug55HLXzrlepO6QyiA3MiOeT515jVMYcOhvld+XjGiGAeFtvD2a0IFPSaENwZQQvCMDmiRBAY\\u003d\",\"tag\":\"Lv5GX9dq2axzKagRML8S1FrU6/T4hGhrB7sv714NCmw\\u003d\"}"#.to_string(),
+                        signature: r#"MEUCIHV5hJ3ijgF+pOdqviqX6NYoIOKdAQvLOevFDW+WF2iGAiEA1vnaeat8ZoSuIp8V6o2aiPcC9TdDyeR0RvImCY8vQWs\u003d"#.to_string(),
+                        version: "ECv1".to_string()
```
Comment (Narayanbhat166) [inline crates/router/src/connector/fiserv/transformers.rs:143]: Please remove the hardcoded data

File: crates/router/src/connector/fiserv/transformers.rs [hunk 5]
```diff
@@ -147,6 +187,47 @@ impl TryFrom<&types::ConnectorAuthType> for FiservAuthType {
...
+                merchant_id: auth.merchant_account,
+                terminal_id: Some(session.terminal_id),
+            },
+            reference_transaction_details: ReferenceTransactionDetails {
+                reference_transaction_id: item.request.connector_transaction_id.to_string(),
+            },
+            transaction_details: TransactionDetails {
+                capture_flag: None,
+                // reversal_reason_code:Some(ReversalReasonCodes::try_from(cancellation_reason)?),
+                reversal_reason_code: Some(cancellation_reason.parse_enum("ReversalReasonCodes").change_context(errors::ConnectorError::MissingRequiredField { field_name: "cancellation reason should be: VOID, TIMEOUT, SUSPECTED_FRAUD, CARD_OVERRIDE" })?)
```
Comment (Narayanbhat166) [inline crates/router/src/connector/fiserv/transformers.rs:225]: Raise `MissingRequiredField` only if data is missing. If the provided data cannot be parsed, raise `InvalidData` error, if it is not present in connector error add it and propogate it.

File: crates/router/src/connector/fiserv/transformers.rs [hunk 6]
```diff
@@ -64,60 +90,74 @@ pub struct TransactionInteraction {
...
 
+        let merchant_details = MerchantDetails {
+            merchant_id: auth.merchant_account,
+            terminal_id: Some(session.terminal_id),
+        };
+
+        let transaction_interaction = TransactionInteraction {
+            origin: "ECOM".to_string(), //Payment is being made in online mode, card not present
+            eci_indicator: "CHANNEL_ENCRYPTED".to_string(), // transaction encryption such as SSL/TLS, but authentication was not performed
+            pos_condition_code: "CARD_NOT_PRESENT_ECOM".to_string(), //card not present in online transaction
```
Comment (Narayanbhat166) [inline crates/router/src/connector/fiserv/transformers.rs:121]: Use enums and explain the enum variants with comments, this will be helpful in extending for other transactionInteraction types.

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/fiserv/transformers.rs [hunk 7]
```diff
@@ -288,56 +434,147 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for FiservCaptureRequest {
...
+#[serde(rename_all = "camelCase")]
+pub struct RefundResponse {
+    gateway_response: GatewayResponse,
+}
+
+#[derive(Default, Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+#[serde(transparent)]
+pub struct RefundSyncResponse {
+    vector_of_sync_response: Vec<FiservPaymentsResponse>,
```
Comment (NishantJoshi00) [inline crates/router/src/connector/fiserv/transformers.rs:529]: Maybe use something like `sync_responses`

File: crates/router/src/connector/fiserv/transformers.rs [hunk 8]
```diff
@@ -188,12 +269,31 @@ impl From<FiservPaymentStatus> for enums::AttemptStatus {
...
 #[serde(rename_all = "camelCase")]
 pub struct FiservPaymentsResponse {
     gateway_response: GatewayResponse,
 }
 
+#[derive(Default, Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+#[serde(transparent)]
+pub struct FiservPaymentsSyncResponse {
+    vector_of_sync_response: Vec<FiservPaymentsResponse>,
```
Comment (NishantJoshi00) [inline crates/router/src/connector/fiserv/transformers.rs:294]: Maybe use something like `sync_responses`

File: crates/router/src/connector/fiserv/transformers.rs [hunk 9]
```diff
@@ -64,60 +90,74 @@ pub struct TransactionInteraction {
...
-                    transaction_interaction,
-                })
+                Source::PaymentCard { card }
+            }
+            api::PaymentMethod::Wallet(wallet_data) => {
+                // let json_value = wallet_data.token
+                // .ok_or_else(utils::missing_field_err("Wallet token"))?
+                // .parse_struct("GooglePayToken")
+                // .change_context(errors::ConnectorError::RequestEncodingFailed)?;
+                // println!("oooooooooo{:?}", json_value);
```
Comment (NishantJoshi00) [inline crates/router/src/connector/fiserv/transformers.rs:138]: Consider removing this, if not required

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/fiserv.rs (hunk 3 above)
Comment (Narayanbhat166) [reply crates/router/src/connector/fiserv.rs:267]: https://github.com/juspay/hyperswitch/blob/8d2e573a29671bbbf74b7646b87c409b25d1b171/crates/router/src/lib.rs#L44

File: crates/router/src/connector/fiserv/transformers.rs [hunk 10]
```diff
@@ -18,11 +19,19 @@ pub struct FiservPaymentsRequest {
...
-    source_type: String,
-    card: CardData,
+#[derive(Debug, Serialize, Eq, PartialEq)]
+#[serde(tag = "sourceType")]
+pub enum Source {
+    PaymentCard {
+        card: CardData,
+    },
+    #[allow(dead_code)]
+    //todo() implement google pay for fiserv
```
Comment (NishantJoshi00) [inline crates/router/src/connector/fiserv/transformers.rs:29]: remove unwanted comments. IF required us the proper format.
```rust
// [#{issue_no}]: {short description}
```

File: crates/router/src/connector/fiserv/transformers.rs [hunk 11]
```diff
@@ -287,56 +435,147 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for FiservCaptureRequest {
...
+impl TryFrom<types::RefundsResponseRouterData<api::RSync, RefundSyncResponse>>
     for types::RefundsRouterData<api::RSync>
 {
     type Error = error_stack::Report<errors::ConnectorError>;
     fn try_from(
-        _item: types::RefundsResponseRouterData<api::RSync, RefundResponse>,
+        item: types::RefundsResponseRouterData<api::RSync, RefundSyncResponse>,
     ) -> Result<Self, Self::Error> {
-        Err(errors::ConnectorError::NotImplemented("fiserv".to_string()).into())
+        let gateway_resp = match item.response.sync_responses.first() {
```
Comment (NishantJoshi00) [inline crates/router/src/connector/fiserv/transformers.rs:561]: Just a suggestion, consider using `first().ok_or(...)`

File: crates/router/src/connector/fiserv.rs [hunk 12]
```diff
@@ -49,7 +50,46 @@ impl Fiserv {
...
+{
+    fn build_headers(
+        &self,
+        req: &types::RouterData<Flow, Request, Response>,
+        _connectors: &settings::Connectors,
+    ) -> CustomResult<Vec<(String, String)>, errors::ConnectorError> {
+        let timestamp = OffsetDateTime::now_utc().unix_timestamp_nanos() / 1_000_000;
+        let auth: fiserv::FiservAuthType =
+            fiserv::FiservAuthType::try_from(&req.connector_auth_type)?;
+        // let api_key = auth.api_key.clone();
```
Comment (NishantJoshi00) [inline crates/router/src/connector/fiserv.rs:65]: remove this comment, if not needed

File: crates/router/src/connector/fiserv/transformers.rs [hunk 13]
```diff
@@ -43,81 +59,108 @@ pub struct Amount {
...
+        let merchant_details = MerchantDetails {
+            merchant_id: auth.merchant_account,
+            terminal_id: Some(session.terminal_id),
+        };
+
+        let transaction_interaction = TransactionInteraction {
+            origin: TransactionInteractionOrigin::Ecom, //Payment is being made in online mode, card not present
+            eci_indicator: TransactionInteractionEciIndicator::ChannelEncrypted, // transaction encryption such as SSL/TLS, but authentication was not performed
+            pos_condition_code: TransactionInteractionPosConditionCode::CardNotPresentEcom, //card not present in online transaction
+        };
```
Comment (NishantJoshi00) [inline crates/router/src/connector/fiserv/transformers.rs:142]:
```suggestion
        let transaction_interaction = TransactionInteraction {
            // Payment is being made in online mode, card not present
            origin: TransactionInteractionOrigin::Ecom,
            // transaction encryption such as SSL/TLS, but authentication was not performed
            eci_indicator: TransactionInteractionEciIndicator::ChannelEncrypted,
            // card not present in online transaction
            pos_condition_code: TransactionInteractionPosConditionCode::CardNotPresentEcom, 
        };
```

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]
Other than that, looks good to me

File: crates/router/src/connector/fiserv/transformers.rs [hunk 14]
```diff
@@ -147,6 +193,46 @@ impl TryFrom<&types::ConnectorAuthType> for FiservAuthType {
...
+}
+
+impl TryFrom<&types::PaymentsCancelRouterData> for FiservCancelRequest {
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(item: &types::PaymentsCancelRouterData) -> Result<Self, Self::Error> {
+        let auth: FiservAuthType = FiservAuthType::try_from(&item.connector_auth_type)?;
+        let metadata = item
+            .connector_meta_data
+            .clone()
+            .ok_or(errors::ConnectorError::RequestEncodingFailed)?;
```
Comment (jagan-jaya) [inline crates/router/src/connector/fiserv/transformers.rs:211]:
```suggestion
        let metadata = item.get_connector_meta_data()?;
```
use it from utils

File: crates/router/src/connector/fiserv/transformers.rs [hunk 15]
```diff
@@ -147,6 +193,46 @@ impl TryFrom<&types::ConnectorAuthType> for FiservAuthType {
...
+        let session: SessionObject = metadata
+            .parse_value("SessionObject")
+            .change_context(errors::ConnectorError::RequestEncodingFailed)?;
```
Comment (jagan-jaya) [inline crates/router/src/connector/fiserv/transformers.rs:214]: move this to utils and use it as `metadata.get_session_object()?`

File: crates/router/src/connector/fiserv/transformers.rs [hunk 16]
```diff
@@ -147,6 +193,46 @@ impl TryFrom<&types::ConnectorAuthType> for FiservAuthType {
...
+        let cancellation_reason = item
+            .request
+            .cancellation_reason
+            .clone()
+            .ok_or_else(utils::missing_field_err("cancellation_reason"))?;
```
Comment (jagan-jaya) [inline crates/router/src/connector/fiserv/transformers.rs:219]: move this to utils and use it as item.request.get_cancellation_reason()?

File: crates/router/src/connector/fiserv/transformers.rs [hunk 17]
```diff
@@ -234,6 +339,48 @@ impl<F, T>
...
+            Some(gateway_response) => gateway_response,
+            _ => Err(errors::ConnectorError::ResponseHandlingFailed)?,
+        };
+
+        Ok(Self {
+            status: gateway_resp
+                .gateway_response
+                .transaction_state
+                .clone()
+                .into(),
```
Comment (jagan-jaya) [inline crates/router/src/connector/fiserv/transformers.rs:366]: don't use into, call the appropriate the `from` implementation for it. this is help in debugging

File: crates/router/src/connector/fiserv/transformers.rs [hunk 18]
```diff
@@ -287,56 +437,148 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for FiservCaptureRequest {
...
+            merchant_details: MerchantDetails {
+                merchant_id: auth.merchant_account,
+                terminal_id: None,
+            },
+            reference_transaction_details: ReferenceTransactionDetails {
+                reference_transaction_id: item
+                    .request
+                    .connector_transaction_id
+                    .get_connector_transaction_id()
+                    .change_context(errors::ConnectorError::MissingConnectorTransactionID)?,
```
Comment (jagan-jaya) [inline crates/router/src/connector/fiserv/transformers.rs:461]: try moving this to connector/utils.rs , this is duplicated in many places

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/fiserv/transformers.rs (hunk 15 above)
Comment (SamraatBansal) [reply crates/router/src/connector/fiserv/transformers.rs:214]: This is a connector specific custom `session Object` created in transformers itself, it cannot go under utils

**Review by Narayanbhat166 — APPROVED** [review]

File: crates/router/src/connector/fiserv/transformers.rs [hunk 19]
```diff
@@ -18,11 +19,18 @@ pub struct FiservPaymentsRequest {
...
+pub enum Source {
+    PaymentCard {
+        card: CardData,
+    },
+    #[allow(dead_code)]
+    GooglePay {
+        data: String,
+        signature: String,
+        version: String,
+    },
```
Comment (NishantJoshi00) [inline crates/router/src/connector/fiserv/transformers.rs:33]: consider removing the `#[allow(dead_code)]` or the `GooglePay` varient, you can do this is a separate PR if need be.

File: crates/router/src/connector/fiserv.rs [hunk 20]
```diff
@@ -61,16 +99,53 @@ impl api::ConnectorCommon for Fiserv {
...
+            (Some(err), _) => err
+                .into_iter()
+                .map(|v| v.message)
+                .collect::<Vec<_>>()
+                .join(""),
+            (None, Some(err_details)) => err_details
+                .into_iter()
+                .map(|v| v.message)
+                .collect::<Vec<_>>()
+                .join(""),
```
Comment (NishantJoshi00) [inline crates/router/src/connector/fiserv.rs:132]:
```suggestion
            (Some(err), _) => err.into_iter().map(|v| v.message).collect::<String>(),
            (None, Some(err_details)) => err_details
                .into_iter()
                .map(|v| v.message)
                .collect::<String>(),
```

Just a minor optimisation, not required to be added here, but consider this for further use-cases

File: crates/router/src/connector/fiserv.rs [hunk 21]
```diff
@@ -401,15 +560,158 @@ impl api::RefundExecute for Fiserv {}
...
+        res: types::Response,
+    ) -> CustomResult<types::RefundSyncRouterData, errors::ConnectorError> {
+        logger::debug!(target: "router::connector::fiserv", response=?res);
+
+        let response: fiserv::RefundSyncResponse = res
+            .response
+            .parse_struct("Fiserv Refund Response")
+            .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
+        types::ResponseRouterData {
+            // response:response.first().get_required_value("gatewayResponse").change_context(errors::ConnectorError::ResponseDeserializationFailed).try_into(),
```
Comment (NishantJoshi00) [inline crates/router/src/connector/fiserv.rs:700]: consider removing this comments, if no longer required

**Review by NishantJoshi00 — APPROVED** [review]
LGTM

**Review by jagan-jaya — APPROVED** [review]

**Review by NishantJoshi00 — DISMISSED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

File: crates/router/src/connector/fiserv/transformers.rs [hunk 22]
```diff
@@ -147,6 +194,39 @@ impl TryFrom<&types::ConnectorAuthType> for FiservAuthType {
...
+            merchant_details: MerchantDetails {
+                merchant_id: auth.merchant_account,
+                terminal_id: Some(session.terminal_id),
+            },
+            reference_transaction_details: ReferenceTransactionDetails {
+                reference_transaction_id: item.request.connector_transaction_id.to_string(),
+            },
+            transaction_details: TransactionDetails {
+                capture_flag: None,
+                reversal_reason_code: Some(cancellation_reason.parse_enum("ReversalReasonCodes").change_context(errors::ConnectorError::InvalidDataFormat { field_name: "cancellation reason should be: VOID, TIMEOUT, SUSPECTED_FRAUD, CARD_OVERRIDE" })?)
```
Comment (Narayanbhat166) [inline crates/router/src/connector/fiserv/transformers.rs:224]: calculate the cancellation_reason outside.

**Review by Narayanbhat166 — COMMENTED** [review]
Minor changes requested

**Comment by jagan-jaya** [comment]
@SamraatBansal can we fix the review comments and close this PR

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by jagan-jaya — DISMISSED** [review]

File: crates/router/src/connector/fiserv/transformers.rs [hunk 23]
```diff
@@ -287,56 +419,150 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for FiservCaptureRequest {
...
+#[serde(rename_all = "camelCase")]
+pub struct RefundResponse {
+    gateway_response: GatewayResponse,
+}
+
+#[derive(Default, Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+#[serde(transparent)]
+pub struct RefundSyncResponse {
+    sync_responses: Vec<FiservPaymentsResponse>,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/fiserv/transformers.rs:516]: If both Psync and RSync response are same keep one struct as "SyncResponse" and implement try from `ResponseRouterData`  and `RefundsResponseRouterData `

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by jarnura — APPROVED** [review]
